File: 프로그래머스/unrated/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.HashMap;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        HashMap<String, Friend> map = new HashMap<>();

        for(int i=0; i<friends.length; i++) map.put(friends[i], new Friend());

        for(int i=0; i<gifts.length; i++) {
            String gift[] = gifts[i].split(" ");

            Friend a = map.get(gift[0]);
            Friend b = map.get(gift[1]);

            a.totalGive += 1;
            a.giveLog.put(gift[1], a.giveLog.getOrDefault(gift[1], 0) + 1);

            b.totalReceive += 1;
            b.receiveLog.put(gift[0], b.receiveLog.getOrDefault(gift[0], 0) + 1);

            map.put(gift[0], a);
            map.put(gift[1], b);
        }

        HashMap<String, Integer> nextMonth = new HashMap<>();
        for(String key: map.keySet()) nextMonth.put(key, 0);

        for(int i=0; i<friends.length-1; i++) {
            Friend a = map.get(friends[i]);

            for(int j=i+1; j< friends.length; j++) {
                Friend b = map.get(friends[j]);

                if(a.giveLog.getOrDefault(friends[j], 0) > a.receiveLog.getOrDefault(friends[j], 0)) {
                    nextMonth.put(friends[i], nextMonth.getOrDefault(friends[i], 0) + 1);
                }else if((a.giveLog.getOrDefault(friends[j], 0) < a.receiveLog.getOrDefault(friends[j], 0))) {
                    nextMonth.put(friends[j], nextMonth.getOrDefault(friends[j], 0) + 1);
                }else {
                   int aGiftScore = a.totalGive - a.totalReceive;
                   int bGiftScore = b.totalGive - b.totalReceive;

                   if(a.giveLog.getOrDefault(friends[j], 0) - a.receiveLog.getOrDefault(friends[j], 0) == 0) {
                       if(bGiftScore > aGiftScore)  nextMonth.put(friends[j], nextMonth.getOrDefault(friends[j], 0) + 1);
                       else if(bGiftScore < aGiftScore) nextMonth.put(friends[i], nextMonth.getOrDefault(friends[i], 0) + 1);
                   }
                }
            }
        }

        for(String key: nextMonth.keySet()) {
            answer = Math.max(nextMonth.getOrDefault(key, 0), answer);
        }


        return answer;
    }
}

class Friend {
    int totalGive = 0;
    int totalReceive = 0;
    HashMap<String, Integer> giveLog = new HashMap<>();
    HashMap<String, Integer> receiveLog = new HashMap<>();
}

File: 프로그래머스/unrated/258712. 가장 많이 받은 선물/가장 많이 받은 선물.kt
import java.util.HashMap

class Solution {
    fun solution(friends: Array<String>, gifts: Array<String>): Int {
        val map = HashMap<String, Friend>()
        friends.forEach { friend -> map[friend] = Friend() }

        gifts.forEach { gift ->
            val parts = gift.split(" ")
            val giver = map[parts[0]]!!
            val receiver = map[parts[1]]!!

            giver.totalGive++
            giver.giveLog[parts[1]] = giver.giveLog.getOrDefault(parts[1], 0) + 1

            receiver.totalReceive++
            receiver.receiveLog[parts[0]] = receiver.receiveLog.getOrDefault(parts[0], 0) + 1
        }

        val nextMonth = HashMap<String, Int>()
        friends.forEach { friend -> nextMonth[friend] = 0 }

        for (i in 0 until friends.size - 1) {
            val a = map[friends[i]]!!

            for (j in i + 1 until friends.size) {
                val b = map[friends[j]]!!

                val aGive = a.giveLog.getOrDefault(friends[j], 0)
                val aReceive = a.receiveLog.getOrDefault(friends[j], 0)
                val aGiftScore = a.totalGive - a.totalReceive
                val bGiftScore = b.totalGive - b.totalReceive

                when {
                    aGive > aReceive -> nextMonth[friends[i]] = nextMonth.getOrDefault(friends[i], 0) + 1
                    aGive < aReceive -> nextMonth[friends[j]] = nextMonth.getOrDefault(friends[j], 0) + 1
                    bGiftScore > aGiftScore -> nextMonth[friends[j]] = nextMonth.getOrDefault(friends[j], 0) + 1
                    bGiftScore < aGiftScore -> nextMonth[friends[i]] = nextMonth.getOrDefault(friends[i], 0) + 1
                }
            }
        }

        return nextMonth.values.maxOrNull() ?: 0
    }
}

class Friend {
    var totalGive = 0
    var totalReceive = 0
    val giveLog = HashMap<String, Int>()
    val receiveLog = HashMap<String, Int>()
}